### PR TITLE
UCL-523 DataGrid search

### PR DESCRIPTION
**Pull request by szmejk**

### CODE REVIEW

File: src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx [hunk 1]
```diff
@@ -52,6 +59,30 @@ const DataGridCellComponent: ForwardRefRenderFunction<HTMLTableCellElement, Prop
     cellStyle.paddingRight = spacing?.paddingRight;
   }
 
+  //NOTE: we might want to migrate to Highlight API once it's supported by Firefox
+  const renderContent = () => {
+    if (typeof children === "string" && searchValue) {
+      if (!children.toLowerCase().includes(searchValue.toLowerCase())) return children;
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx:65]: lack of `{        }`

File: src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx [hunk 2]
```diff
@@ -52,6 +59,30 @@ const DataGridCellComponent: ForwardRefRenderFunction<HTMLTableCellElement, Prop
...
+
+      const matchingSequence = new RegExp(`${searchValue}`, "i").exec(children);
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx:67]:
```suggestion
      const matchingSequence = new RegExp(searchValue, "i").exec(children);
```

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  inputProps?: InputWrapperProps;
+}
+
+export const DataGridSearchbar = ({
+  onSearch,
+  searchValue,
+  debounceTime,
+  inputProps
+}: DataGridSearchbarProps) => {
+  const [search, setSearch] = useState(searchValue ?? "");
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx:36]: if you are letting ppl set `searchValue` would be nice to have that sync, other wise it's only `initialSearchValue`, right :)?
Maybe useEffect to update the internal search state?

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    setSearch(event.target.value);
+  };
+
+  return (
+    <InputWrapper
+      {...inputProps}
+      className={classes["searchbar"]}
+      label={inputProps?.label ?? ""}
+      onChange={onChange}
+      type={"search"}
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx:51]: `type="search"`

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    onSearchCallback(event.target.value);
+    setSearch(event.target.value);
+  };
+
+  return (
+    <InputWrapper
+      {...inputProps}
+      className={classes["searchbar"]}
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx:48]:
```
className={`${classes["searchbar"]} ${inputProps.className}`}
```

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      label={inputProps?.label ?? ""}
+      onChange={onChange}
+      type={"search"}
+      name={inputProps?.name ?? "searchbar"}
+      value={search}
+      inputProps={{
+        ...inputProps?.inputProps,
+        type: "search",
+        prefix: <Icon icon={Icons.Search} />,
+        placeholder: inputProps?.inputProps?.placeholder ?? "Search items"
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx:58]: maybe that placeholder should be a part of main props?

File: src/components/DataGrid/DataGridFilters/DataGridToolbar.module.scss [hunk 7]
```diff
@@ -25,6 +25,10 @@
   width: 100%;
   padding-bottom: 1rem;
 
+  .searchbar {
+    min-width: 15rem;
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridFilters/DataGridToolbar.module.scss:29]: maybe: ```
min-width: 7rem;
flex-basis: 15rem;
```
7rem still 'visible', and 15rem as standard width :)

File: src/components/DataGrid/DataGridFilters/DataGridToolbar.tsx [hunk 8]
```diff
@@ -62,26 +62,19 @@ const filtersReducer = (state: FiltersState, action: FiltersAction): FiltersStat
...
+  onFilterAdd,
+  onFilterEdit,
+  onFilterDelete,
+  onFiltersClear
+}: DataGridToolbarProps) => {
   const [state, dispatch] = useReducer(filtersReducer, { filters: filterValues || [] });
   const { clearButtonCaption = "Clear all filters" } = translations?.toolbar || {};
   return (
-    <div {...rest} ref={ref} className={classes["toolbar"]}>
+    <>
```
Comment (pawel-napieracz) [inline src/components/DataGrid/DataGridFilters/DataGridToolbar.tsx:77]: I like `<>` but we agreed on using `<Fragment>` :P

File: stories/DataGrid/DataGrid.mdx [hunk 9]
```diff
@@ -209,7 +209,7 @@ Text will be displayed only if the `data` property is empty, and `isLoading` is
 
 # Filters section
 
-To show filter toolbar one has to pass `filters` prop to the `DataGrid` component with `enableFilters` property set to `true`.
+To show filter toolbar one has to pass `filters` prop to the `DataGrid` component with `enable` property set to `true`.
```
Comment (pawel-napieracz) [inline stories/DataGrid/DataGrid.mdx:212]: consider to remove `enable` and flat props.

File: stories/DataGrid/DataGrid.mdx [hunk 10]
```diff
@@ -241,6 +241,30 @@ User can configure the behavior of the toolbar using the `filterProps` object wh
...
 
+# Search section
+
+To show searchbar one has to pass `search` prop to the `DataGrid` component with `enable` property set to `true`.
+
+User can configure the behavior of the search using the `searchProps` object which has the following properties:
+
+- `onSearch` - callback function called when user types in searchbar
+- `debounceTime` - time in `ms` used to debounce the `onSearch` callback. When not specified the callback will fire on every `onChange` event
+- `searchValue` - default search value
```
Comment (pawel-napieracz) [inline stories/DataGrid/DataGrid.mdx:252]: `initialSearchValue`

File: stories/DataGrid/DataGrid.stories.tsx [hunk 11]
```diff
@@ -611,3 +611,77 @@ DataGridWithFiltersInEditMode.play = conditionalPlay(async ({ canvasElement }) =
...
+          )}
+        </DataGridComponent>
+      </div>
+    </div>
+  );
+};
+
+export const DataGridWithSearch = SearchTemplate.bind({});
+
+DataGridWithSearch.args = {
```
Comment (pawel-napieracz) [inline stories/DataGrid/DataGrid.stories.tsx:649]: 1. create example with filters and search.
2. Add filtering data on searching to both examples

File: src/components/DataGrid/DataGrid.tsx [hunk 12]
```diff
@@ -48,6 +50,10 @@ export interface Props<T> extends Omit<ComponentPropsWithRef<"div">, "children">
     enable: boolean;
     filtersProps: DataGridToolbarProps;
   };
+  search?: {
+    enable: boolean;
```
[the remark this answers is not in the document]
Comment (szmejk) [reply src/components/DataGrid/DataGrid.tsx]: done

File: src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx (hunk 1 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx]: done

File: src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx (hunk 2 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridCell/DataGridCell.tsx]: done

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx (hunk 3 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx]: done

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx (hunk 4 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx]: done

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx (hunk 5 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx]: done

File: src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx (hunk 6 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridFilters/DataGridSearchbar.tsx]: done

File: src/components/DataGrid/DataGridFilters/DataGridToolbar.module.scss (hunk 7 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridFilters/DataGridToolbar.module.scss]: done

File: stories/DataGrid/DataGrid.mdx (hunk 9 above)
Comment (szmejk) [reply stories/DataGrid/DataGrid.mdx]: done

File: stories/DataGrid/DataGrid.mdx (hunk 10 above)
Comment (szmejk) [reply stories/DataGrid/DataGrid.mdx]: done

File: stories/DataGrid/DataGrid.stories.tsx (hunk 11 above)
Comment (szmejk) [reply stories/DataGrid/DataGrid.stories.tsx:652]: Added filtering on search, I'll create a separate ticket to include all of the filtering options in one toolbar

File: src/components/DataGrid/DataGridFilters/DataGridToolbar.tsx (hunk 8 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridFilters/DataGridToolbar.tsx]: done

**Review by pawel-napieracz — APPROVED** [review]

**Pull request merged**